crates/ide-assists/src/handlers/extract_function.rs

```diff
@@ -1311,37 +1347,144 @@ fn format_function(
     let const_kw = if fun.mods.is_const { "const " } else { "" };
     let async_kw = if fun.control_flow.is_async { "async " } else { "" };
     let unsafe_kw = if fun.control_flow.is_unsafe { "unsafe " } else { "" };
+    let (generic_params, where_clause) = make_generic_params_and_where_clause(ctx, fun);
     match ctx.config.snippet_cap {
         Some(_) => format_to!(
             fn_def,
-            "\n\n{}{}{}{}fn $0{}{}",
+            "\n\n{}{}{}{}fn $0{}",
             new_indent,
             const_kw,
             async_kw,
             unsafe_kw,
             fun.name,
-            params
         ),
         None => format_to!(
             fn_def,
-            "\n\n{}{}{}{}fn {}{}",
+            "\n\n{}{}{}{}fn {}",
             new_indent,
             const_kw,
             async_kw,
             unsafe_kw,
             fun.name,
-            params
         ),
     }
+
+    if let Some(generic_params) = generic_params {
+        format_to!(fn_def, "{}", generic_params);
+    }
+
+    format_to!(fn_def, "{}", params);
+
+    if let Some(where_clause) = where_clause {
+        format_to!(fn_def, " {}", where_clause);
+    }
+
     if let Some(ret_ty) = ret_ty {
         format_to!(fn_def, " {}", ret_ty);
     }
+
     format_to!(fn_def, " {}", body);
 
     fn_def
 }
 
+fn make_generic_params_and_where_clause(
+    ctx: &AssistContext,
+    fun: &Function,
+) -> (Option<ast::GenericParamList>, Option<ast::WhereClause>) {
+    let used_type_params = fun.type_params(ctx);
+
+    let generic_param_list = fun
+        .mods
+        .generic_param_list
+        .as_ref()
+        .map(|parent_params| make_generic_param_list(ctx, parent_params, &used_type_params))
+        .flatten();
+
+    let where_clause =
+        fun.mods.where_clause.as_ref().map(|parent_where_clause| {
+            make_where_clause(ctx, parent_where_clause, &used_type_params)
+        });
+
+    (generic_param_list, where_clause)
+}
+
+fn make_generic_param_list(
+    ctx: &AssistContext,
+    parent_params: &ast::GenericParamList,
+    used_type_params: &[TypeParam],
+) -> Option<ast::GenericParamList> {
+    let required_generic_params: Vec<ast::GenericParam> = parent_params
+        .generic_params()
+        .filter(|param| param_is_required(ctx, param, used_type_params))
+        .collect();
+    if required_generic_params.is_empty() {
+        None
+    } else {
+        Some(make::generic_param_list(required_generic_params))
+    }
+}
+
+fn param_is_required(
+    ctx: &AssistContext,
+    param: &ast::GenericParam,
+    used_type_params: &[TypeParam],
+) -> bool {
+    match param {
+        ast::GenericParam::ConstParam(_) | ast::GenericParam::LifetimeParam(_) => true,
```

Comment:
@Veykril Interested in your thoughts here. I could imagine three potential approaches:
- Always copy `ConstParam` and `LifetimeParam` (as implemented here)
- Never copy `ConstParam` or `LifetimeParam` (essentially a more constrained version of the status quo before this change)
- "Do the right thing" ie filter out unused `ConstParam` and `LifetimeParam`, only include the ones referenced within the function


I feel like "Always Copy" is probably the wrong approach, for basically the same reasons you gave on my initial change.

Never copying these params would be trivial to implement, but potentially frustrating for users.

In order to "do the right thing", I'll need to do a similar kind of filtration as I've already done for `TypeParam`s, for both consts and lifetimes. I think lifetimes should be relatively straightforward, but I'm not exactly sure how I'd go about this for `const` params. For instance, unclear to me when/if they should be passed down as params, or via the turbo fish at the call site.

Reply:
I think I'm currently leaning towards the "never copy" approach, ie changing the `true` above to a `false`, and then handling lifetime and const params in separate future changes.

Reply:
Especially since generic lifetime param support would be limited in usefulness until https://github.com/rust-lang/rust-analyzer/issues/9945 is fixed